### ISSUE TITLE
Frequent disconnection during ssh connection

### ISSUE THREAD

**Issue by munsta0** [FreeRDP/Remmina#909]
Hey all,

Me and a coworker have been experiencing a lot of disconnect recently, mostly in waves. We can go days without incident, then suddenly we disconnect every few minutes, which is _very_ annoying.

We suspect it could be related to poor internet quality. Is there a way to make the program less sensitive to packet loss? Attempt a reconnect instead of just closing?

Thanks
- Client (OS name and version): Ubuntu 14.04
- Remmina version (remmina --version): remmina - Version 1.2.0-rcgit.11
- Desktop environment (GNOME, Unity, KDE, ..): Unity
- Connecting to (OS and version): Centos 5.0
- Connecting via (RDP, VNC, ...): ssh
##

**Comment by antenore** [comment]
The SSH connection is handled through libssh, I'll check if they support a reconnect mechanism  .

**Comment by antenore** [comment]
So it seems there is not a solution for this issue, but just some work around and settings that may help you.

1) Server Side in the /etc/ssh/sshd.conf

Increase the ClientAliveInterval to 60 seconds and set TCPKeepAlive to yes.

2) On the client (server) side

On the server run a _screen_ or _tmux_ session so even if you loose your ssh session, you'll be able to _attach_ to it.

I doubt we can do anything else, but I'll keep the issue opened for investigations some days.

**Comment by munsta0** [comment]
Thanks for the quick reply. I'll check those settings and thanks for the screen program.

**Comment by antenore** [comment]
OK, let me know

**Comment by munsta0** [comment]
Alright changed the settings, in a few days I'll update with the results.

Also, for anyone else having this issue, the file is actually /etc/ssh/sshd_config on Centos 5.0

**Comment by antenore** [comment]
> Also, for anyone else having this issue, the file is actually /etc/ssh/sshd_config on Centos 5.0

My bad, it should be that name everywhere :-P sorry. Except in most *BSD like OS where it's under /usr/local/etc/ssh

**Comment by zeronug** [comment]
This is happening to me on Fedora 25 Workstation.  Super annoying, terminal box just closes.  SSh-ing into Ubuntu 16.10 and Keepalive is 'yes'.

**Comment by zagr0** [comment]
the same for me on Fedora 25. The ssh window/session of remmina just closing. Happens very often.

**Comment by byjg** [comment]
It seems it is an old bug in Remmina. It is not related to SSH environment. I do the follow tests:

- Open a SSH connection inside a Gnome Terminal -> The connection remains from several minutes without disconnect even if I do not touch in the keyboard;
- Open a Terminal in Remmina to the same server above -> The connection dies and the terminal just hang if I do not touch on the keyboard for 1 minute or less;

I created a working around -> If i have to keep the remmina terminal opened and connected I just do a infinite ping to some server inside the remmina terminal. The connection in remmina continues open and do not hang. For this reason I believe this is a bug in Remmina or it is "By Design". 

I know this "behavior" for more than 3 years.

**Comment by thiagomgo** [comment]
It seems remmina doesn't respect the `~/.ssh/config` file and on Ubuntu 16.04, it creates a `know_hosts` file in `~/snap/remmina/common/.ssh/`. Even if I add a `/config` file into this folder, it is ignored. I'm using remmina 1.2.0-rcgit-24,

**Comment by antenore** [comment]
It's not Remmina but the way snap works as it's isolated from the rest of the operating system. 
It's a feature....

**Comment by byjg** [comment]
My Remmina installation does not run under the snap environment. I installed directly through "apt". 

This problem is totally ignored by the remmina developers. Because of this I migrated and now I use  "Guake" + "Guake Indicator".

**Comment by thiagomgo** [comment]
Thanks @byjg, I didn't know this feature. I'll take a look.

**Comment by antenore** [comment]
It's not totally ignored by us  @byjg , I don't have solutions at the moment.
Your behavior doesn't help that much in finding the cause and fixing it definitively. This is just trolling.

**Comment by byjg** [comment]
Sorry, that's not my intention.  Comment edited. There are only the relevant information about "snap" environment. 

Unfortunately I am not strong enough in C/Glade to dig into the code. 

My contribution is that for some reason Remmina does not honor the config in /etc/ssh/ssh_config. In the past I did some tests. When my /etc/ssh/ssh_config is not set properly (ServerAliveInterval and ServerAliveMaxCount), even gnome-terminal hangs after a short inactivity time. After I setup this, gnome-terminal works as expected, but remmina dont. 

I do not if it make sense but  I hope it help you.

**Comment by antenore** [comment]
Unlucky to solve the issue I've too few options and none of them are quick to be implemented.

**Comment by antenore** [comment]
So, there was a misunderstanding from my side of the issue (sorry).
@giox069 have implemented the keepalive in Remmina more than [a year ago](https://github.com/FreeRDP/Remmina/commit/56dc6c729ae06f6f7bd9568d00bc7a3dab4263e4). 
It looks like on some distros/versions the connection is kept alive.

Can you please all give me you kernel keep alive parameters?

```sh
  # cat /proc/sys/net/ipv4/tcp_keepalive_time
  # cat /proc/sys/net/ipv4/tcp_keepalive_intvl
  # cat /proc/sys/net/ipv4/tcp_keepalive_probes
```
These parameters can be overridden in the application forcing them to other values, that it's what have been done by Giovanni.

Now, why it works on some distro and not on other it's quite a mystery.

**Comment by byjg** [comment]
No problem. I confirm that issue persists with me. My installed version is Remmina 1.1.2. I installed from APT package:  `remmina 1.1.2-3ubuntu1 amd64`. My system is:

```
$ lsb_release --all
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial
```
and here is what you ask to me:

```
$ cat /proc/sys/net/ipv4/tcp_keepalive_time
7200
$ cat /proc/sys/net/ipv4/tcp_keepalive_intvl
75
$ cat /proc/sys/net/ipv4/tcp_keepalive_probes
9
```

**Comment by giox069** [comment]
@byjg remmina 1.1.2 is extremely old. To install a newer remmina on ubuntu 16.04 use our PPA as explained in the [Remmina Wiki](https://github.com/FreeRDP/Remmina/wiki). The last version of remmina is 1.2.0-rcgit.24.
But I can confirm you that the ssh disconnection problem is still here on 1.2.0-rcgit.24: keepalive packets are sent every 5 seconds, but there is a global timeout of 15 minutes of I/O inactivity that whe should find and destroy :)

**Comment by cambohost** [comment]
I'm going to chime in here and mention that this bug is making Remmina un-usable. Some observations:

- Constant SSH disconnections, even over a VPN directly to the server I'm working with. 
- Keep-alive settings all tweaked on both ends to no avail
- Connections to CentOS machines get dropped much quicker than Ubuntu machines
- Sometimes the disconnects occur right in the middle of typing something
- Disconnects do not occur with sessions to localhost
- Nothing in the logs
- Connections to the same machine over the same VPN with the same settings using any other terminal client remain connected for days/weeks at a time without issue

I'm using Ubuntu 17.10 artful.

**Comment by antenore** [comment]
Thanks for these additional information @cambohost 

The SSH Remmina plugin doesn't yet have a mechanism to recover from connection losses, while RDP for example has this functionality.

I could possibly find how to do the same for the SSH protocol, but I was wondering and procrastinating about the fact that this should be implemented by libssh, the library we use behind Remmina, it could be also that a way to recover from a connection lost, it's already available with libssh, and we haven't implemented it.

It's quite complex and I have an hard time finding the information I need.

Any help it's much appreciated.

In the meanwhile it's very important to understand why you have connections issues and try as much as possible to fix or avoid those problems.

I pass through VPNs, firewalls with authentications and proxies where I work, and except when I'm on wifi and move in the building, I never encounter this problem. I've SSH connections opened for several hours often and the only way I've to replicate you problems, it's to temporary disconnect the rj45 or disable the wifi connection.

**Comment by antenore** [comment]
I've committed some code into next that may improve the ssh connection, I've tested it on Arch Linux and it seems to work properly.

Anybody available to compile from source code and let me know? In the wiki you have detailed instructions for most of the distros.

**Comment by zagr0** [comment]
Looks ok now, thnx for the fix!
fedora 27, remmina 1.2.0-rcgit-27 (git 3a6cde40)

**Comment by antenore** [comment]
Resolved via #1553
